### ISSUE TITLE
Setting voltage, Turning power on and getting a different voltage then I set?

### ISSUE THREAD

**Issue by geekbozu** [kanflo/opendps#17]
Not sure if this is an actual bug. But I got my DPS today flashed it with OpenDPS and was playing with it. I currently set the voltage to 5.00 and when I turn it on it drops down to 4.74 even with out a load attached? Is this normal/expected behavior?

**Comment by kanflo** [comment]
@geekbozu That would be unexpected behavior. Is 4.74V consistent between the display and when measured with a multimeter? Did you verify the functionality of the DPS with the original firmware?

@fromeijn Could you please explain "open loop system"?

**Comment by fromeijn** [comment]
With "open loop system" I mean, that the dac is set without reading it back. There is no PID controller of some sort between it.

**Comment by geekbozu** [comment]
From my measurements it was accurate. I will not lie how ever I did not
verify before flashing I made sure an led lit up then flashed away. I did
save the constants for 5V-On and OFF however.If it is any consolation the
measurement of the input voltage also matches my power supply I am using.
It does seems to be in agreement with what @fromeijn is saying. As the
error is less drastic at lower voltages and gets worse as I get higher up.

**Comment by kanflo** [comment]
Thanks for the explanation @fromeijn. A PID controller would be very useful, I will create a new issue to track it.

**Comment by geekbozu** [comment]
So I quadruple checked. It seems that the on screen display is correct give or take .02v according to my multi-meter.  Would using a feedback look make more sense then say a per unit calibration? Would almost seem like its worth making a manual calibration function to tune each unit that way. (and to have power levels not have to adjust as much after being turned on)

**Comment by kanflo** [comment]
A PID controller would be better IMHO as people have observed voltage drops at higher loads.

**Comment by geekbozu** [comment]
I think I was not clear. I more meant would a way to calibrate be useful In addition to the PID controller. Or do the devices seem accurate enough from the factory that the PID tuning should be quick enough to not matter. I know I have burned components up due to power supplies that jump to higher voltages on startup before settling down. So that is where my concern is from there.

**Comment by kanflo** [comment]
Hmmm, you have a point there. The controller would have to be extremely quick to react to reductions in load as not to fry things. Calibration could be a better option for those not using higher loads.

**Comment by geekbozu** [comment]
Well the thought is you need both. One to handle initial startup/Hardware
discrepancies, The PID to handle variations in Load + Drift. That being
said I am more then willing to work on a calibration interface/setup if
need be once the CC branch is merged (no point in starting another project
on a split codebase IMO) and could probably tinker with the PID tuning if
@fromeijn does not jump on that.

**Comment by Manawyrm** [comment]
Can confirm, with voltage set to 5.00V, it outputs 4.729V (and displays 4.71V)...
I'd also say that a calibration/offset setting would probably be better than a PID loop, which could overshoot and kill sensitive things...

**Comment by fromeijn** [comment]
Good suggestion, we don't want any magic smoke. So we need a calibration routine for voltage and current.

**Comment by geekbozu** [comment]
I guess I will start familiarizing my self with the code base. :D Seems like a worthwhile effort to me.

**Comment by vogelfreiheit** [comment]
Did anyone get somewhere with either self-calibration or PID? I have updated calibration values for my 5015. Both DAC and ADC.

**Comment by geekbozu** [comment]
I have been waiting for CC mode to be implemented to not fracture the code
more before I started working on it, I do plan on implementing it soon that
being said.

**Comment by geekbozu** [comment]
So over time it seems my device has also lost its factory calibration. Im reading 7.8v when I supply it 5v.

**Comment by kanflo** [comment]
For OpenDPS devices there is no factory calibration but it seems a calibration function just got pushed up the TODO list.

**Comment by geekbozu** [comment]
I will dig into it a bit more once I have some time my self, and as noted I'll gladly tackle setting up some kind of calibration UI/scheme once the CC branch is "merged"

**Comment by kanflo** [comment]
Adding a calibration UI scheme would be quite easy now as the UI supports different screens with editable widgets. See [func_cv.c](https://github.com/kanflo/opendps/blob/master/opendps/func_cv.c) and [func_cc.c](https://github.com/kanflo/opendps/blob/master/opendps/func_cc.c).

**Comment by antonblanchard** [comment]
My DPS5005 exhibited the same issue. Before flashing I had taken dumps at a few voltage levels, so I could work out the correct V_DAC_K and V_DAC_C settings:

```
#define V_DAC_K (double)0.0776
#define V_DAC_C (double)2.0
```

My mulitmeter, while not calibrated, showed voltages were spot on after this change. I then adjusted V_ADC_K so that the voltages displayed matched:

```
#define V_ADC_K (double)12.8891
```

I tested the open circuit voltage as well as under almost 5A of load, and the voltage was consistent. This suggests the issue is just a calibration one, and a PID is not required.

**Comment by vogelfreiheit** [comment]
Late to the party, I will go through the comments.... glad other people are still working on this. I'm still working towards a successful dump of the original firmware :>

**Comment by Robin10155** [comment]
Can somebody please explain what does V_DAC_C and V_DAC_K means? I am trying to calibrate the DPS5005. Currently it is off by 300mV at 5V.

**Comment by kanflo** [comment]
@Robin10155 It's the equation of the straight line describing the voltage output as a function of the DAC setting:

```
V_out = DAC_setting * V_DAC_K + V_DAC_C
```

See [this spreadsheet](https://docs.google.com/spreadsheets/d/1AhGsU_gvZjqZyr2ZYrnkz6BeUqMquzh9UNYoTqy_Zp4/edit?usp=sharing). You can calibrate your unit by measuring V_out given a series of DAC settings and plotting the values.

**Comment by Robin10155** [comment]
Thanks for a quick reply. I will give it a slot.

**Comment by kanflo** [comment]
From #48: "I see exactly the same behaviour wrt. output voltage on my new '5005. I will look into solving this at build time."

**Comment by Xenoamor** [comment]
Thanks. If you find what is different on the schematic let us know!

**Comment by geekbozu** [comment]
Could, At your leisure, Explain how to use the PAST library you have included as well as clarify the bug that is noted in your comments? I plan to get to this soon (finally) and want to make sure that we can save the configs to start >.>

With that would getting a command line based calibrator make more sense first? Or should we go for from panel interface-able calibration right off the bat?

**Comment by geekbozu** [comment]
Yeah that is what I was thinking instead of writing on on hardware ui to start. 

Process something like such.

"Set Dac to value" -> User types in real value
x2-3 times
Once voltage is done, 
Ask user for Resistor value, -> set output voltage and measure current @ known value
x2-3 times

Save and have a well calibrated output. 

Input calibration will be a little trickier but not horrid.
Working on this now starting with getting it to load constants from past then we can extend dpsctl.

**Comment by vogelfreiheit** [comment]
I second both the PID controller loop idea and also having an UI "debug menu" of sorts (text screen is plenty good) that we could activate and do the calibration there. Adding a requirement for dpsctl means another layer of complexity and extra hardware when we could literally just do this with a barebones UI that takes Vin and Vout and perhaps Cin Cout and adjust things internally.

Nobody should use the firmware AS-IS with the built-in calibration constants because there is no guarantee whatsoever of those being accurate. These values are directly tied to the hardware. Every board will differ. The original firmware uses a PID loop if I am not mistaken.

**Comment by geekbozu** [comment]
I already have my local copy loading constants from flash, I'm currently working on implementing a way to send the constants via DPSCTL once that's functional I will be working on a UI wizard for it. Because I agree it needs something. 

Once we have the ability to know out constants work across the voltage and amperage range we can/will work on some kind of run time adjustment.

**Comment by vogelfreiheit** [comment]
Unrelated but: we could add MQTT capabilities to the ESP32 firmware. Not sure what is available and the code quality involved, but that would make it ridiculously useful for integration with all kinds of automation systems and metrics gathering.   If some issues are ironed out, the whole package would rival (based on cost alone) pretty much anything else the average hobbyist could get his hands on.

**Comment by geekbozu** [comment]
Throw it in the  SCPI support issue both would be pretty sweet

**Comment by geekbozu** [comment]
Hit a small snag where ATOF is using more memory then originally intended so I need to rework how dpsctl sends commands for calibration over. But progress is still happening. Slowly in my FORK

**Comment by Xenoamor** [comment]
Are there any downsides to moving from fixed point to floating point, I believe there's no floating point unit in the MCU?

Also I think there are a few lines like this: https://github.com/geekbozu/opendps/blob/c1edaaac687a703266e88f62cda0f17f033ca8cd/opendps/pwrctl.c#L275 that will lose the fractional part (i.e. 1.9 will become 1)

**Comment by geekbozu** [comment]
IMO not really it will be a bit slower but not really enough to matter for what we are doing. OCP is the only thing that will be really effected by it and OCP should be set to be more forgiving anyway then @ death point of the Device Under Load. 

For the items like what was linked is not really ok but acceptable as depending on the case.  

It is a value being sent to the dac and needs to be a whole number anyway.
or
Is a value with a resolution of .01mV, The floating point is to help with rounding/accuracy down to a few uV. 

However they probably should be rounded instead of truncated so your "closer" to your real target. I'll adjust it later on if needed Not particularly relevant for this issue. Once Calibration is good we can automate some data logging to see how off it may be and if we need to make adjustments :)

**Comment by Xenoamor** [comment]
I believe Kanflo pre-calculates the OCP value with pwrctl_calc_ilimit_adc() and stores it in the variable pwrctl_i_limit_raw so speed shouldn't be an issue when it comes to this

**Comment by vogelfreiheit** [comment]
How is this progressing?

**Comment by geekbozu** [comment]
Coming back around on the round robin soon. I need to make a "ui" still and that is really it for calibration you can manually specify the values if you know what they are with my version of DPSCTL.py (supposing I pushed the right version there was a small bug last I checked). Making a case for it then its back to making this work right so i'm not developing it on my lap.

**Comment by geekbozu** [comment]
https://imgur.com/a/WwTwlj2
Welp case is made. Really happy with it so now I can sanely get back to this. Hopefully I get some free time from work soon.

Note: Line 120 Protocol.py needs to be "f" not "d" for anyone testing will be fixed in the next commit.

**Comment by geekbozu** [comment]
https://imgur.com/a/WwTwlj2
Welp case is made. Really happy with it so now I can sanely get back to this. Hopefully I get some free time from work soon.
Is it possible to update the boot loader with out the SWD wire-up? I really don't want to attempt to solder that backup any time soon if I don't have to. 

Note: Line 120 in my fork Protocol.py needs to be "f" not "d" for anyone testing will be fixed in the next commit.

**Comment by geekbozu** [comment]
So I have gotten Output Voltage calibration working 👍 
Input Voltage Calibration is next. 
Current/Constant current mode are after that. 

If anyone wants to test I would be grateful.
https://github.com/geekbozu/opendps/commit/46df2c9d939d98602ddfeae8ec2095003c535fc8

code for dpsctl for calibration
`dpsctl.py -d comX -c`

I added a re-init function incase your DPS gets really messed up, It reverts back to the constants supplied in dps-model.h the code for that is
`dpsctl.py -d comX --init`

Thanks!
[Testing Binary](http://img.keepdream.in/opendps.bin)

**Comment by geekbozu** [comment]
@kanflo is the formula for calibrating the V_IN ADC different then V_OUT? I cant seem to produce results that match real world using the current formulas, Nor do your spread sheets match the values supplied in pwrctl.c

**Comment by Xenoamor** [comment]
This is great! Thanks so much. I'll be sure to test this as soon as I'm back from an international work trip

**Comment by geekbozu** [comment]
Kinda, For input voltage calibration we need to get a reference point on the ADC, We do that by feeding a known voltage to the DPS we can get known ADC values to compute our coefficients. 

I am currently still debating on how I want to handle current calibration, I might have a way to handle both CC and Current display in one shot. I'm looking over if it will work well for us.  :)

**Comment by kanflo** [comment]
Really cool work @geekbozu, hoping to try this out tomorrow!

**Comment by vogelfreiheit** [comment]
I'll try this over the weekend! I might need to look into soldering a ESP32 directly to the UART and have some sort of bridge or such for this purpose. Anyone has any custom firmware they recommend? It could be useful to have graphing or data output from the PSU (maybe fake SCPI?).

**Comment by geekbozu** [comment]
It is not wired up in the pictures but in mine I got a USB bulkhead adapter and wired up a USB -> Serial adapter for mine. https://imgur.com/a/WwTwlj2 You can see the bulkhead in the last picture.

**Comment by Xenoamor** [comment]
You might want an isolated USB to serial adaptor

**Comment by geekbozu** [comment]
Only ground is wired up, So it may introduce some noise to the circuit/get some noise from the circuit but other then that it should be fine, I honestly don't plan on using it with USB plugged in in most cases anyway its really just for development and occasional data logging.

**Comment by kanflo** [comment]
I finally got around to test and I am impressed! Much looking forward to your PR.

**Comment by geekbozu** [comment]
I actually have been waiting for someone other then me to test this as I haven't had much time to put into it! So its been sitting on my backburner waiting for voltage calibration to be tested before I start working on current calibration!

**Comment by Xenoamor** [comment]
Will test now

**Comment by geekbozu** [comment]
I will gladly accept any input on how to make this program flow better. User side software is definitely not my strong point :)

**Comment by Xenoamor** [comment]
Ooops... disregard that I haven't had my evening coffee. I input the values out by a factor of 10....
It seems to have corrected the 140mV offset that the device had on the input with the default values so all is working well!

I'll try the rest now

**Comment by geekbozu** [comment]
Woo :) Definitely need to spruce up the instructions but for now "they work"

**Comment by Xenoamor** [comment]
It seems to be working to me. I think with some code cleanup and some clearer instructions in the python tools this should be good to go. I'll start making pushes to your repo for these

**Comment by geekbozu** [comment]
That would be amazing, Thank you a ton. Ill work on tackling Current calibration soonish Should have time later this week.

**Comment by kanflo** [comment]
If you store the coefficients in the past they will survive power cycles. I see you store the K and C values in different units which in theory may cause problems with atomicity. I would recommend storing them in the same unit.

**Comment by geekbozu** [comment]
I can add the mangling to store them into 1 past unit later. Due to the fact that there is only one place where they are written and read they should always reflect the last written value.

**Comment by geekbozu** [comment]
It very much is possible to re-work that, I am not super versed on how to effectively handle blocks of code like that so I defaulted back to If/Else. Thankfully it does work in its current state and at the least should not need to be expanded to much Honestly this code needs more love IMO https://github.com/geekbozu/opendps/blob/a8d7248c63a8b69777929b5a6959c5ae0520c9af/opendps/pwrctl.c#L55


Definitely agree those routines are a mess in both the Python file and the C source code. They work but they are not really beautiful code 👍

**Comment by Xenoamor** [comment]
I didn't realise that the serial commands operate on `variable=value`. Please disregard my comments on changing [opendps_set_calibration](https://github.com/geekbozu/opendps/blob/1da1c1f8eab3db899b42696d3ccb18ea233ee3d4/opendps/opendps.c#L274) to use enums. I'll look at cleaning up handle_set_calibration/handle_list_parameters

**Comment by Xenoamor** [comment]
I've noticed a strange issue with when you calibrate the output voltage

```
$ sudo python dpsctl.py -d /dev/ttyUSB0 -c
To restore the device to factory default calibration please use dpsctl.py --init

Please ensure nothing is connected to the output of the DPS before starting calibration!

Perform Input Voltage Calibration? (y/n): n
Perform Output Voltage Calibration? (y/n): y
You will need an accurate method of measuring voltage, such as a multimeter

Proceed (y/n): y
DPS supply voltage in mV: 16000
Cal Point 1, 10% of Max
voltage: ok
current: ok
Measured voltage on output in mV: 1543
Cal Point 2, 90% of Max
voltage: ok
Measured voltage on output in mV: 13720
['V_DAC_K=0.0756343927076', 'V_DAC_C=0.29613205223', 'V_ADC_K=13.2214983713', 'V_ADC_C=-83.2442996743']
Unknown response 15 from device.
Output Voltage Calibration Complete
Perform Output Current Calibration? (y/n): n
Perform Constant Current Calibration? (y/n): n
$ sudo python dpsctl.py -d /dev/ttyUSB0 -cr
Calibration Report:
        A_ADC_K = 1.71300005913
        A_ADC_C = -118.510002136
        A_DAC_K = 0.652000010014
        A_DAC_C = 288.610992432
        V_ADC_K = 0.0756343901157
        V_ADC_C = 0.296132057905
        V_DAC_K = 0.0756343901157
        V_DAC_C = 0.296132057905
        VIN_ADC_K = 16.7714881897
        VIN_ADC_C = 201.257858276
        VIN_ADC = 942
        VOUT_ADC = 7
        IOUT_ADC = 70
        IOUT_DAC = 0
        VOUT_DAC = 0
```

It seems like `V_ADC_K` is being set to the value of `V_DAC_K` and `V_ADC_C` is being set to the value of `V_DAC_C`

**Comment by Xenoamor** [comment]
Right I _believe_ this is all good to go now. If anyone gets a chance to test it [the repo can be found here](https://github.com/Xenoamor/opendps)

It's fixed the issues I had on my DPS5005 which were all way out before I calibrated it

**Comment by geekbozu** [comment]
I've been following, I just havent had time. Life and holidays are reeking havoc with my schedule

**Comment by geekbozu** [comment]
I looked at your repo, I was planning on doing current calibration differently, Instead of needing multiple resistors once we know that Output is correct we can use one nominal resistor value for the whole range of available current. With out any user interaction. Should be more user friendly to just give a measured Low value resistor and handle it that way.

**Comment by Xenoamor** [comment]
That seems like a good way of doing it. I guess you could first calibrate the current ADC and then straight after calibrate the constant current mode automatically

**Comment by geekbozu** [comment]
that was the thought, It requires a strict in order calibration process. Input -> output -> current. But I feel its less burdensome on the user. It not like this is a nist traceable calibration after all :)
Any other opinions @kanflo ?

**Comment by Xenoamor** [comment]
Isn't it already important that you do the output voltage calibration before you do the current calibration though as it works out the current from V/R? If that is the case then I don't see there being a much greater hinderance from having to do the input voltage first

**Comment by geekbozu** [comment]
Well the way it is programmed now its very important. We could in theory have the user generate all the values. But then what would be the point of calibration to begin with ;) Its also very important todo input calibration before all of those else we cant properly set a min/max operating voltage!

The current code for current calibration was a copy paste and honestly if it works I'm shocked FWIW. It was me setting up boilerplate for organization reasons. Bad habits die hard I know.

**Comment by kanflo** [comment]
@geekbozu I think "strict order calibration process" beats "burdensome for the user" ;) I will have a look at your work this Christmas.

**Comment by geekbozu** [comment]
Glad we are in agreement! And merry Xmas Everyone!

**Comment by Xenoamor** [comment]
Okay the calibration routine has been changed to have a fixed order. It got my device down to under 1% accuracy from ~5%. My multimeter doesn't have many counts so a better one would likely get it lower

The way I did it is the following:
1. Remove everything from the output of the DPS5005
2. Power the DPS5005 from a bench power supply set to ~7V
3. Run `python dpsctl.py -d COMX --init`. If you have wildly wrong values calibrated you get crazy results so returning to the defaults is a good idea
4. Run `python dpsctl.py -d COMX -c`
5. Multimeter the input terminals and input the value read in mV
6. Turn up the bench power supply to ~15V
7. Multimeter the input terminals and input the value read in mV
8. Multimeter the output terminals and input the value read in mV
9. Multimeter the output terminals again and input the value read in mV
10. Enter the max output current of your DPS device (e.g. 5000mA for the DPS5005)
11. Enter your load resistor in ohms
12. Proceed if your load resistor can take the required amperage
13. The device will now iterate through 10 voltages and use this to calibrate the current ADC

Below is the output from this
```
python dpsctl.py -d COM7 --init

python dpsctl.py -d COM7 -c
For calibration you will need:
        A multimeter
        A known load capable of handling the required power
        2 stable input voltages

To restore the device to factory default calibration please use dpsctl.py --init

Please ensure nothing is connected to the output of the DPS before starting calibration!

Input Voltage Calibration:

Please hook up the first lower supply voltage to the DPS now
ensuring that the serial connection is connected after boot
Type input voltage in mV: 7090

Please hook up the second higher supply voltage to the DPS now
ensuring that the serial connection is connected after boot
Type input voltage in mV: 15030
Input Voltage Calibration Complete

Output Voltage Calibration:
Cal Point 1, 10% of Max
voltage: ok
current: ok
Measured voltage on output in mV: 1443
Cal Point 2, 90% of Max
voltage: ok
current: ok
Measured voltage on output in mV: 12910
Output Voltage Calibration Complete

Output Current Calibration:
DPS max Amperage in mA: 5000
Load resistance in ohms: 14.1
Load must be rated for at least 16.0 watts!
Please connect load to the output of the DPS, then press enter
voltage: ok
current: ok
voltage: ok
current: ok
voltage: ok
current: ok
voltage: ok
current: ok
voltage: ok
current: ok
voltage: ok
current: ok
voltage: ok
current: ok
voltage: ok
current: ok
voltage: ok
current: ok
Output Current Calibration Complete

Constant Current Calibration:
NOT IMPLEMENTED
```

I don't really understand how the constant current stuff works so I've left this unimplemented. Perhaps if someone could explain it to me I'll add it in?

The test repo can be found here: https://github.com/Xenoamor/opendps

**Comment by Xenoamor** [comment]
Okay I've now added constant current calibration (although it doesn't seem to be as accurate as I would like)

I've also made it so it restores whatever settings you had setup before the calibration routine so it doesn't wipe them. It also is a lot less verbose now so you won't get the "current: ok" prompts

Let me know if anyone needs anymore details so they can test it. This seems all good to go on my side

**Comment by Xenoamor** [comment]
Okay I've been using this for a while now and it seems okay. However one of the parameters, A_ADC_C, is being corrupted on a power cycle. Could this be a bug in past.c? I don't think there's anything wrong in my code additions/changes..
```
python dpsctl.py -d COM7 -cr
Calibration Report:
        A_ADC_K = 1.70325279236
        A_ADC_C = -115.116912842
        A_DAC_K = 0.662139832973
        A_DAC_C = 238.729721069
        V_ADC_K = 13.2220907211
        V_ADC_C = -84.8741073608
        V_DAC_K = 0.0756310075521
        V_DAC_C = 0.419114351273
        VIN_ADC_K = 16.8350524902
        VIN_ADC_C = 59.6804122925
        VIN_ADC = 913
        VOUT_ADC = 7
        IOUT_ADC = 70
        IOUT_DAC = 0
        VOUT_DAC = 0

-- Power cycle the DPS5005

python dpsctl.py -d COM7 -cr
Calibration Report:
        A_ADC_K = 1.70325279236
        A_ADC_C = 1.0732764899e-08 <-- this is now erroneous
        A_DAC_K = 0.662139832973
        A_DAC_C = 238.729721069
        V_ADC_K = 13.2220907211
        V_ADC_C = -84.8741073608
        V_DAC_K = 0.0756310075521
        V_DAC_C = 0.419114351273
        VIN_ADC_K = 16.8350524902
        VIN_ADC_C = 59.6804122925
        VIN_ADC = 913
        VOUT_ADC = 7
        IOUT_ADC = 0
        IOUT_DAC = 0
        VOUT_DAC = 0
```

**Comment by geekbozu** [comment]
i'm honestly kinda stumped on this one for the moment, I see nothign glaring in the code. Seriously though thanks for all the help!
Could you do some more power cycles and see if its just that param that corrupts or if others do eventually as well?

**Comment by Xenoamor** [comment]
The mystery deepens... it appears it sets the A_ADC_C value to this strange value even if I run a --init. Perhaps it's not being erased properly?
```
python dpsctl.py -d COM7 --init

python dpsctl.py -d COM7 -cr
Calibration Report:
        A_ADC_K = 1.71300005913
        A_ADC_C = -118.510002136
        A_DAC_K = 0.652000010014
        A_DAC_C = 288.610992432
        V_ADC_K = 13.1639995575
        V_ADC_C = -100.750999451
        V_DAC_K = 0.0719999969006
        V_DAC_C = 1.85000002384
        VIN_ADC_K = 16.7460002899
        VIN_ADC_C = 64.1119995117
        VIN_ADC = 913
        VOUT_ADC = 7
        IOUT_ADC = 0
        IOUT_DAC = 0
        VOUT_DAC = 0

-- Power cycle the DPS5005

python dpsctl.py -d COM7 -cr
Calibration Report:
        A_ADC_K = 1.71300005913
        A_ADC_C = 1.0732764899e-08 <-- this is now erroneous
        A_DAC_K = 0.652000010014
        A_DAC_C = 288.610992432
        V_ADC_K = 13.1639995575
        V_ADC_C = -100.750999451
        V_DAC_K = 0.0719999969006
        V_DAC_C = 1.85000002384
        VIN_ADC_K = 16.7460002899
        VIN_ADC_C = 64.1119995117
        VIN_ADC = 914
        VOUT_ADC = 6
        IOUT_ADC = 0
        IOUT_DAC = 0
        VOUT_DAC = 0
```

**Comment by Xenoamor** [comment]
Okay so I changed the enum values in pastunits.h and it's now working...
```
typedef enum {
    /** stored as [I_limit:16] | [V_out:16] */
    past_power = 1,
    cal_A_ADC_K = 5, <--- now starting from 5
    cal_A_ADC_C,
    cal_A_DAC_K,
    cal_A_DAC_C,
    cal_V_DAC_K,
    cal_V_DAC_C,
    cal_V_ADC_K,
    cal_V_ADC_C,
    cal_VIN_ADC_K,
    cal_VIN_ADC_C,
    /** stored as 0 or 1 */
    past_tft_inversion,
    /** stored as strings */
    past_boot_git_hash,
    past_app_git_hash,
    /** A past unit who's precense indicates we have a non finished upgrade and
    must not boot */
    past_upgrade_started = 0xff
} parameter_id_t;
```

@kanflo do you know what might be happening here? Is it possible I have a really b0rked past flash section that isn't being erased?

If anyone can test it with the standard code and confirm this is a "me" problem then that would be great

**Comment by Xenoamor** [comment]
Okay the "1.0732764899e-08" erroneous value is stuck in the ID space 3. Is something else using this past ID?

**Comment by Xenoamor** [comment]
Okay I've found the offending line here: https://github.com/kanflo/opendps/blob/2e1a8e686c8f3b910c00cb737a68af8efa043e21/dpsboot/dpsboot.c#L308

Looks like DPSboot is changing it. Now there's nothing wrong with this but I'll make it clear in pastunits.h and I'll also remove 'past_power = 1' as this is now not used

**Comment by Xenoamor** [comment]
[Commit has now been made](https://github.com/Xenoamor/opendps/commit/d9d243755058a0846fc88e5a26a23a13e9e1c953) and I believe we're looking to be able to merge this in again

**Comment by geekbozu** [comment]
Your amazing. I spent my time checking the CV/CC modes for the same bug, Never thought to look in the boot-loader. Ill build and flash this tonight and give it a check. Really excited thanks for all the help!!!!!

**Comment by Xenoamor** [comment]
It could do with taking multiple ADC readings and averaging them to be honest but my python-fu isn't very good so I'm not too sure how to do this without cluttering the codebase

**Comment by kanflo** [comment]
I probably should add a comment regarding the past unit IDs stating that they must not be shuffled around. Glad it wasn't a bug ;)

**Comment by geekbozu** [comment]
Once this has been proven to work, I can take a stab at the python side. Sadly nothing will be pretty with this but I'm sure there is a way to clean up the code and make it "prettier" we shall see. 

I think a better comment to be made would be to ensure that all pastId's are properly documented in said enum. however I think that is likely hard seeing as the boot-loader and opendps do have "seperate" code bases. Either way I'm glad it was caught as well, thinking back I think I had this issue as well and was not sure if I was just going crazy or not! probably why I wanted someone else to test so bad.

**Comment by geekbozu** [comment]
So I've tested it. We need to add a way for the DPS to ignore current overage during calibration. My unit refuses to calibrate up because it immediately overloads upon starting and it gets bad readings as a result.

**Comment by Xenoamor** [comment]
I'm actually thinking about removing the current overage completely in #65. I too had the same issue though on occasion. The issue you're referring to is the same as seen in #6

**Comment by geekbozu** [comment]
I did manage to get it to work after disabling OCP and grabbing a 510Ω resistor. The constant current routine needs some love still My results not consistent. However I feel this may be to the voltage sag associated with the lack of PID or what ever we are missing there. Then the in accuracy of the routines.

I personally agree that #65 is the way it should be as well FWIW

**Comment by Xenoamor** [comment]
The only known voltage sag is due to the issue in #7 which won't be fixed by a PID. This is because the ADC value for the voltage is effectively wrong so there's no way to adjust for this until that issue is resolved. Once you have calibrated the open circuit output voltage using this routine you shouldn't see any discrepancy between the voltage set on the display and the voltage that is read back. When you read this with a multimeter though you will notice that this voltage is actually wrong under high current loads as a voltage is dropped across the shunt resistor. This in turn means it will be calculating/calibrating the current incorrectly at the top end

Perhaps we should calibrate it by having a multimeter in series with the device and do it much the same as we do the output voltage calibration? I'm less comfortable doing this though as most multimeters have low value fuses on some inputs

Saying all this though my calibrated device is now within 10mA in constant current mode where it used to be out by over 100mA:
![image](https://user-images.githubusercontent.com/7109951/50653085-4c30ae00-0f80-11e9-83b2-b22f94c4236d.png)
The top meter is reading 1.00A and the bottom meter is reading 14.347V (this is lower because of #7)

**Comment by Xenoamor** [comment]
Heres the DPS5005 in CV mode with an output of 10V set:
![image](https://user-images.githubusercontent.com/7109951/50653435-6ae37480-0f81-11e9-975e-89f85e086a57.png)
The top meter is reading 0.70A and the bottom meter is reading 9.997V

**Comment by Xenoamor** [comment]
Okay I've been thinking about this and I think what we need is a way of directly setting the DAC values and then working out the calibration coefficients from this directly. This current method relies on the default coefficients being in the right ballpark which is not great if it's a new model or a new hardware build.

This would be tricky however as you'd have to tell the calibration routine when the device tops out and use that as the maximum. This could be done by reducing the DAC from the maximum in 10% steps until the user sees the voltage drop

I believe it would also be better to have the current calibration work done by putting a multimeter in series. Now you'd have to attach it to the high amperage input on the meter to prevent any accidental fuses blowing but I think this is likely to be a more accurate method of doing it then by using a resistive load

**Comment by geekbozu** [comment]
I see no reason why the resistive load would not be "accurate" afaik that is the NIST standard, Just using HIGH power High tolerance resistors. 

I agree having it handle DAC values would be better. Definitely going to ponder this! I think I would like CC/CV modes merged first however. I feel like that holds this unit back quite a bit. 

also apologies for like vanishing. Life has gotten a bit hectic I'll be back to this soon! Thank you for all the help in the mean time!

**Comment by Xenoamor** [comment]
That's a good point. I think I'm worrying needlessly and when it worked for me it worked well.

I'll work on the DAC stuff which I've now labelled as #103 and then get an input and output voltage calibration routine implemented. I'd merge CC/CV modes like you said but I'd rather sort this out now for #100.

Don't worry about being around too much, we're all working on this for fun so feel free to drop in and out as you please!

**Comment by ajdalshov** [comment]
> My DPS5005 exhibited the same issue. Before flashing I had taken dumps at a few voltage levels, so I could work out the correct V_DAC_K and V_DAC_C settings:

Interesting. Could you please elaborate on how you got the values from the dumps? I'd love to do the same with my dumps I made prior to flashing. Thanks. :)

**Comment by Ha644** [comment]
Hallo.
I flashed dps5005 with opendps firmware. 
There is an issue I’m working in constant current mode when I set the current at 4ma and measuring the output it will give 4.084 on multimeter as will as on dps5005 screen.
Can anyone help how to subtract this offset and get the 0ma as output as well.
